src/schema.rs

```diff
@@ -0,0 +1,68 @@
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("unknown type {0}, choose one of {1:?}")]
+    UnknownType(String, &'static [&'static str]),
+    #[error("error generating JSON: {0}")]
+    GenerateJson(#[from] serde_json::Error),
+}
+
+use schemars::{
+    gen::SchemaSettings,
+    visit::{self, Visitor},
+};
+use std::str::FromStr;
+
+pub struct Schema(pub schemars::schema::RootSchema);
+
+impl TryFrom<crate::curr::TypeVariant> for Schema {
+    type Error = Error;
+    fn try_from(variant: crate::curr::TypeVariant) -> Result<Schema, Error> {
+        println!("hhh");
+        let settings = SchemaSettings::draft07().with_visitor(ReplaceAdditionalProperties);
+        let generator = settings.into_generator();
+        Ok(Schema(variant.json_schema(generator)))
+    }
+}
+
+impl FromStr for Schema {
+    type Err = Error;
+    fn from_str(s: &str) -> Result<Schema, Error> {
+        s.parse::<crate::curr::TypeVariant>()
+            .map_err(|_| {
+                Error::UnknownType(s.to_string(), &crate::curr::TypeVariant::VARIANTS_STR)
+            })?
+            .try_into()
+    }
+}
```

Comment:
What's depending on the TryFrom and the FromStr?

I'm not sure how we can support different versions of schema with them, and it looks like they primarily exist to inject the ReplaceAdditionalProperties which we should really do at the generation of the schema in xdrgen.

We'd also need these to be embedded under curr and next.